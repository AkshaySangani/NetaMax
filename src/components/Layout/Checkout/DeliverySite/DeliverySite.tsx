import { ReactElement } from "react";

import { Box, Flex, Img, Link, Text } from "@chakra-ui/react";

import { IDeliverySiteProps } from "./IDeliverySiteProps";

/**
 * The DeliverySite component.
 * @param { IDeliverySiteProps } props the component props.
 * @returns { ReactElement } The DeliverySite component.
 */
export const DeliverySite = (props: IDeliverySiteProps): ReactElement => {
  const { selectedStore, updateStore, fromSuccess } = props;

  return (
    <Box
      h="auto"
      width="calc(100% + 3rem)"
      marginLeft="-1.5rem"
      bg="#f6f6f6"
      display="flex"
      alignItems="center"
    >
      <Box p="16px" display="flex" flexDirection="column">
        <Text
          color="#000000"
          fontSize="20px"
          fontWeight="700"
          display="flex"
          alignItems="flex-start"
        >
          Sitio de entrega
        </Text>
        <Text fontSize="16px" color="#5f5f5f">
          Esta es la tiendita que elegiste para la entrega
        </Text>
        <Flex marginTop="16px">
          <Box width="calc(30%)" padding="10px">
            <Flex
              height="60px"
              padding="auto"
              alignItems="center"
              justifyContent="center"
              backgroundColor="#d9d9d9"
              borderRadius="5px"
            >
              <Img width="30px" src="/assets/images/store.png" />
            </Flex>
          </Box>
          <Box padding="8px" width="calc(70% + 3rem)">
            <Text fontWeight="700" fontSize="16px">
              {selectedStore?.name}
            </Text>
            <Text
              marginTop="2px"
              fontSize="14px"
              display="flex"
              alignItems="center"
              color="#5f5f5f"
            >
              {selectedStore?.companyAddress}
            </Text>
            {fromSuccess ? (
              <>
                <Link
                  marginTop="2px"
                  color="#3870FF"
                  fontWeight="700"
                  decoration="underline"
                  target="_blank"
                  href={`https://www.google.com/maps/place/${selectedStore?.location.lat},${selectedStore?.location.lng}`}
                >
                  Ver ubicación
                </Link>
                <Text marginTop="10px" fontWeight="700">
                  Contacta con la tiendita
                </Text>
                <Link
                  marginTop="2px"
                  color="#3870FF"
                  fontWeight="700"
                  decoration="underline"
                  target="_blank"
                  href={`https://wa.me/52${selectedStore?.companyPhoneNumber}`}
                >
                  Ir al whatsapp de la tiendita
                </Link>
              </>
            ) : (
              <Text
                marginTop="2px"
                color="#3870FF"
                fontWeight="700"
                decoration="underline"
                onClick={updateStore}
              >
                Cambiar tiendita
              </Text>
            )}
          </Box>
        </Flex>
      </Box>
    </Box>
  );
};
