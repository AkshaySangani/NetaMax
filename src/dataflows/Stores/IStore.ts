import { ILocation } from "./IStoreState";

export interface IStore {
  /**
   * The unique identifier for the item.
   * @type {string}
   */
  id: string;

  /**
   * The name identifies the item.
   * @type {string}
   */
  name: string;

  /**
   * Is a url to the store
   * @type {string}
   */
  url: string;

  /**
   * Is a host to the store
   * @type {string}
   */
  hosts: string;

  /**
   * Company name.
   * @type {string}
   */
  companyName: string;

  /**
   * Company name.
   * @type {string}
   */
  companyAddress: string;

  /**
   * Company phone number.
   * @type {string}
   */
  companyPhoneNumber: string;
}

export interface IResponse {
  stores: IStoreResponse[];
  distance: number;
}

export interface IStoreResponse {
  /**
   * The stores id
   * @type {string}
   */
  id: string;

  /**
   * The stores name
   * @type {string}
   */
  name: string;

  /**
   * The stores url
   * @type {string}
   */
  url: string;

  /**
   * The stores hosts
   * @type {string}
   */
  hosts: string;

  /**
   * The stores company name
   * @type {string}
   */
  companyName: string;

  /**
   * The stores company address
   * @type {string}
   */
  companyAddress: string;

  /**
   * The stores company phone number
   * @type {string}
   */
  companyPhoneNumber: string;

  /**
   * The stores distance in meters
   * @type {string}
   */
  storeDistance: string;

  /**
   * The stores distance in minutes
   * @type {string}
   */
  storeDistanceInMinutes: string;

  /**
   * The stores latitud
   * @type {string}
   */
  latitud: string;

  /**
   * The stores longitud
   * @type {string}
   */
  longitud: string;

  /**
   * Location coordinates of the store.
   * @type {ILocation}
   */
  location?: ILocation;
}
