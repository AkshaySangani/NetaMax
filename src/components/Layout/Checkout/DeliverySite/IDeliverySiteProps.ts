import { INearestStore } from "dataflows/Stores/INearestStore";
import { IStore } from "dataflows/Stores/IStore";

export interface IDeliverySiteProps {
  /**
   * Allows to change the store selected
   * @type {() => void}
   */
  updateStore?: () => void;

  /**
   * The store selected
   * @type {INearestStore}
   */
  selectedStore?: INearestStore;

  /**
   * Indicates if its called from the order success
   * @type {boolean}
   */
  fromSuccess?: boolean;
}
