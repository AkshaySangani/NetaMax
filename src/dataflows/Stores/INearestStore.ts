import { IStore } from "./IStore";
import { ILocation } from "./IStoreState";

export interface INearestStore extends IStore {
  /**
   * The distance in meters to the store
   * @type {string}
   */
  storeDistance: string;

  /**
   *  The distance in minutes to the store
   * @type {string}
   */
  storeDistanceInMinutes: string;

  /**
   * Location coordinates of the store.
   * @type {ILocation}
   */
  location: ILocation;
}
