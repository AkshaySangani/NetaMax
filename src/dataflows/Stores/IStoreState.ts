import { INearestStore } from "./INearestStore";
import { IStore } from "./IStore";

export interface IStoreState {
  /**
   * The array of stores
   * @type {IStore}
   **/
  store?: IStore;

  /**
   * Indicates if the products are loading.
   * @type {boolean}
   */
  isLoadingStore: boolean;

  /**
   * Indicates if the nearest stores are loading
   * @type {boolean}
   **/
  isLoadingNearestStores: boolean;

  /**
   * Indicates the nearest stores
   * @type {INearestStore[]}
   **/
  nearestStores: INearestStore[];

  /**
   * The error message
   * @type {string}
   */
  errorMessage?: string;

  /**
   * The error message
   * @type {ILocation}
   */
  userLocation?: ILocation;

  /**
   * selected store in map
   * @type {INearestStore}
   */
  selectedStore?: INearestStore;

  /**
   * Distance for the current locations search
   * @type {number}
   */
  circleDistance: number;
}

export interface ILocation {
  /**
   * Locations latitude
   * @type {number}
   */
  lat: number;

  /**
   * Locations longitude
   * @type {number}
   */
  lng: number;

  /**
   * Address name if possible
   * @type {string}
   */
  addressName?: string;
}
