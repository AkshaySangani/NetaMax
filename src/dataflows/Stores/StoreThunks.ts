import { get } from "utils/http";

import { createAsyncThunk } from "@reduxjs/toolkit";

import { INearestStore } from "./INearestStore";
import { IResponse, IStore } from "./IStore";

export const getStoreByName = createAsyncThunk("store/getStoreDetail", async (name: string) => {
  const response = await get(`store?url=${name}`);
  const stores = (await response.data) as IStore[];
  if (Array.isArray(stores) && stores.length !== 0) return stores[0];
  else return undefined;
});

export const getNearestStores = createAsyncThunk(
  "store/getStores",
  async (data: { lat: number; lng: number; search: string }) => {
    const response = await get(
      `store/by/location?location=${encodeURIComponent(data.search)}&lat=${data.lat}&lng=${
        data.lng
      }`
    );
    const typedResponse = response.data as IResponse;
    const nearestStores = typedResponse.stores;
    if (Array.isArray(nearestStores) && nearestStores.length !== 0) {
      nearestStores.map((store) => {
        store.location = { lat: Number(store.latitud), lng: Number(store.longitud) };

        return store;
      });
      return {
        nearestStores: nearestStores as INearestStore[],
        distance: typedResponse.distance,
      };
    } else throw Error("No nearest stores found.");
  }
);
