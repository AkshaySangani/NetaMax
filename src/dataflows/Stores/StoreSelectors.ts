import { RootState } from "state/store";

import { INearestStore } from "./INearestStore";
import { IStore } from "./IStore";
import { ILocation } from "./IStoreState";

/**
 * Selector to get the sales products.
 * @param {RootState} state the root state
 * @returns {IProduct[]} the sales products.
 */
export const selectStore = (state: RootState): IStore | undefined => state?.store?.store;

/**
 * Selector to get the isLoadingSalesSection state from the state.
 * @param {RootState} state the root state
 * @returns {boolean} the isLoadingSalesSection state
 */
export const selectIsLoadingStore = (state: RootState): boolean => state?.store?.isLoadingStore;

/**
 * Selector to get the nearest stores
 * @param {RootState} state the root state
 * @returns {Array} the nearest stores
 */
export const selectNearestStores = (state: RootState): INearestStore[] => state.store.nearestStores;

/**
 * Selector to get if is loading nearest stores
 * @param {RootState} state the root state
 * @returns {boolean} loading nearest stores
 */
export const selectLoadingNearestStores = (state: RootState): boolean =>
  state.store.isLoadingNearestStores;

/**
 * Gets the error message
 * @param {RootState} state the root state
 * @returns {string | undefined} the error message
 */
export const selectErrorMessage = (state: RootState): string | undefined =>
  state.store.errorMessage;

/**
 * Gets the user location
 * @param {RootState} state the root state
 * @returns {ILocation | undefined} the user location
 */
export const selectUserLocation = (state: RootState): ILocation | undefined =>
  state.store.userLocation;

/**
 * Gets the selected store
 * @param {RootState} state the root state
 * @returns {INearestStore | undefined} the selected store
 */
export const selectSelectedStore = (state: RootState): INearestStore | undefined =>
  state.store.selectedStore;

/**
 * Gets the circle distance
 * @param {RootState} state the root state
 * @returns {number} the error message
 */
export const selectCircleDistance = (state: RootState): number => state.store.circleDistance;
