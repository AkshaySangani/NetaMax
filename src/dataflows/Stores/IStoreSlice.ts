import { createSlice, PayloadAction } from "@reduxjs/toolkit";

import { IStore } from "./IStore";
import { IStoreState } from "./IStoreState";
import { getNearestStores, getStoreByName } from "./StoreThunks";

const initialState: IStoreState = {
  store: undefined,
  isLoadingStore: false,
  nearestStores: [],
  isLoadingNearestStores: false,
  circleDistance: 2500,
};

const storeSlice = createSlice({
  name: "store",
  initialState,
  reducers: {
    resetNearestStores: (state) => {
      state.nearestStores = [];
    },
    setStoreAsSelected: (state, action: PayloadAction<IStore>) => {
      state.store = action.payload;
    },
    setUserLocation: (state, action) => {
      state.userLocation = action.payload;
    },
    setSelectedStore: (state, action) => {
      state.selectedStore = action.payload;
    },
    clearSelectedStore: (state) => {
      state.selectedStore = undefined;
    },
    clearErrorMessage: (state) => {
      state.errorMessage = undefined;
    },
  },
  extraReducers: {
    [getStoreByName.pending.type]: (state) => {
      state.isLoadingStore = true;
    },
    [getStoreByName.fulfilled.type]: (state, action) => {
      state.store = action.payload;
      state.isLoadingStore = false;
    },
    [getNearestStores.pending.type]: (state) => {
      state.isLoadingNearestStores = true;
      state.nearestStores = [];
      state.circleDistance = 2500;
      state.errorMessage = undefined;
    },
    [getNearestStores.fulfilled.type]: (state, action) => {
      state.isLoadingNearestStores = false;
      state.nearestStores = action.payload.nearestStores;
      state.circleDistance = action.payload.distance;
      state.errorMessage = undefined;
    },
    [getNearestStores.rejected.type]: (state) => {
      state.isLoadingNearestStores = false;
      state.nearestStores = [];
      state.circleDistance = 2500;
      state.errorMessage =
        "No encontramos tienditas cercanas a la dirección añadida, intenta con otra.";
    },
  },
});

/**
 * Actions
 */
export const {
  resetNearestStores,
  setStoreAsSelected,
  setUserLocation,
  setSelectedStore,
  clearSelectedStore,
  clearErrorMessage,
} = storeSlice.actions;

/**
 * Reducers
 */
export const storeReducer = storeSlice.reducer;
