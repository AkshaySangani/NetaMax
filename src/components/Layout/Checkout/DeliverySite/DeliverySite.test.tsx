import { cleanup, render, screen } from "@testing-library/react";
import { INearestStore } from "dataflows/Stores/INearestStore";
import { DeliverySite } from "./DeliverySite";
import { IDeliverySiteProps } from "./IDeliverySiteProps";

describe("DeliverySite component test", () => {
  let mock: IDeliverySiteProps;
  let selectedStore: INearestStore;
  beforeEach(() => {
    selectedStore = {
      id: "1",
      name: "Tienda Agustin",
      url: "google.com",
      hosts: "google.com",
      companyAddress: "Lamartine 111",
      companyName: "Tienda Agustin",
      companyPhoneNumber: "1111111111",
      storeDistance: "10.000",
      storeDistanceInMinutes: "10.000",
      location: {
        lat: 20,
        lng: 20,
      },
    };
    mock = {
      updateStore: jest.fn(),
      selectedStore,
    };
  });

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  test("DeliverySite should be rendered correctly", () => {
    const { container } = render(<DeliverySite {...mock} />);
    expect(container).toMatchSnapshot();
  });

  test("DeliverySite store name should be in the document", () => {
    render(<DeliverySite {...mock} />);
    expect(screen.getByText(mock.selectedStore?.name || "no store found")).toBeInTheDocument();
  });

  test("DeliverySite store address should be in the document", () => {
    render(<DeliverySite {...mock} />);
    expect(
      screen.getByText(mock.selectedStore?.companyAddress || "no address")
    ).toBeInTheDocument();
  });

  test("DeliverySite cambiar tiendita should be in the document", () => {
    render(<DeliverySite {...mock} />);
    expect(screen.getByText("Cambiar tiendita")).toBeInTheDocument();
  });
});
