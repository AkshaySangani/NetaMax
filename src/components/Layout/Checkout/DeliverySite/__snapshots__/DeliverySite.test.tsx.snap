// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeliverySite component test DeliverySite should be rendered correctly 1`] = `
<div>
  <div
    class="css-k8wr52"
  >
    <div
      class="css-1siwe0b"
    >
      <p
        class="chakra-text css-okz3aw"
      >
        Sitio de entrega
      </p>
      <p
        class="chakra-text css-8whdou"
      >
        Esta es la tiendita que elegiste para la entrega
      </p>
      <div
        class="css-mifb2i"
      >
        <div
          class="css-2jtlu4"
        >
          <div
            class="css-qim7r6"
          >
            <img
              class="chakra-image css-34k09d"
              src="/assets/images/store.png"
            />
          </div>
        </div>
        <div
          class="css-12w7o40"
        >
          <p
            class="chakra-text css-1mzljxq"
          >
            Tienda Agustin
          </p>
          <p
            class="chakra-text css-w6om4s"
          >
            Lamartine 111
          </p>
          <p
            class="chakra-text css-r4awxu"
          >
            Cambiar tiendita
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
